fdbserver/workloads/MiniCycle.actor.cpp

```diff
@@ -0,0 +1,293 @@
+#include "fdbserver/TesterInterface.actor.h"
+#include "fdbserver/workloads/workloads.actor.h"
+#include "fdbserver/workloads/BulkSetup.actor.h"
+#include "flow/Arena.h"
+#include "flow/IRandom.h"
+#include "flow/Trace.h"
+#include "flow/actorcompiler.h" // This must be the last #include.
+#include "flow/serialize.h"
+#include <cstring>
+
+struct MiniCycleWorkload : TestWorkload {
+	int actorCount, nodeCount;
+	double testDuration, transactionsPerSecond, minExpectedTransactionsPerSecond, traceParentProbability;
+	Key keyPrefix;
+
+	std::vector<Future<Void>> clients;
+	PerfIntCounter transactions, retries, tooOldRetries, commitFailedRetries;
+	PerfDoubleCounter totalLatency;
+
+	MiniCycleWorkload(WorkloadContext const& wcx)
+	  : TestWorkload(wcx), transactions("Transactions"), retries("Retries"), tooOldRetries("Retries.too_old"),
+	    commitFailedRetries("Retries.commit_failed"), totalLatency("Latency") {
+		testDuration = getOption(options, "testDuration"_sr, 10.0);
+		transactionsPerSecond = getOption(options, "transactionsPerSecond"_sr, 5000.0) / clientCount;
+		actorCount = getOption(options, "actorsPerClient"_sr, transactionsPerSecond / 5);
+		nodeCount = getOption(options, "nodeCount"_sr, transactionsPerSecond * clientCount);
+		keyPrefix = unprintable(getOption(options, "keyPrefix"_sr, LiteralStringRef("")).toString());
+		traceParentProbability = getOption(options, "traceParentProbability "_sr, 0.01);
+		minExpectedTransactionsPerSecond = transactionsPerSecond * getOption(options, "expectedRate"_sr, 0.7);
+	}
+
+	std::string description() const override { return "MiniCycleWorkload"; }
+	Future<Void> setup(Database const& cx) override { return bulkSetup(cx, this, nodeCount, Promise<double>()); }
+	Future<Void> start(Database const& cx) override {
+		for (int c = 0; c < actorCount; c++)
+			clients.push_back(
+			    timeout(cycleClient(cx->clone(), this, actorCount / transactionsPerSecond), testDuration, Void()));
+		return delay(testDuration);
+	}
+	Future<bool> check(Database const& cx) override {
+		int errors = 0;
+		for (int c = 0; c < clients.size(); c++)
+			errors += clients[c].isError();
+		if (errors)
+			TraceEvent(SevError, "TestFailure").detail("Reason", "There were client errors.");
+		clients.clear();
+		return cycleCheck(cx->clone(), this, !errors);
+	}
+	void getMetrics(std::vector<PerfMetric>& m) override {
+		m.push_back(transactions.getMetric());
+		m.push_back(retries.getMetric());
+		m.push_back(tooOldRetries.getMetric());
+		m.push_back(commitFailedRetries.getMetric());
+		m.emplace_back("Avg Latency (ms)", 1000 * totalLatency.getValue() / transactions.getValue(), Averaged::True);
+		m.emplace_back("Read rows/simsec (approx)", transactions.getValue() * 3 / testDuration, Averaged::False);
+		m.emplace_back("Write rows/simsec (approx)", transactions.getValue() * 4 / testDuration, Averaged::False);
+	}
+
+	Key keyForIndex(int n) { return key(n); }
+	Key key(int n) { return doubleToTestKey((double)n, keyPrefix); }
```

Comment:
This was:
```
Key key(int n) { return doubleToTestKey((double)n / nodeCount, keyPrefix); }
```

I would need to change it to keep n within the mini-cycle and thus in between `[offset, cycleSize + rem + offset)`.  I don't quite understand what's happening here though.